service/league_service/league_rater.py

```diff
@@ -18,4 +18,68 @@ def rate(
         outcome: GameOutcome,
         player_rating: Rating,
     ):
+        boost = 0
+        reduction = 0
+        rating = player_rating[0] - 3 * player_rating[1]
+        player_div = cls._get_player_division(current_score.division_id, league)
+        # Highest divison has lowest id
+        lowest_id, highest_id = cls._get_division_id_range()
+
+        if current_score.game_count >= 10:
+            if current_score.division_id is None:
+                return cls._do_placement(league, current_score, rating)
+
+            if rating > player_div.max_rating:
+                boost = 1
+            elif rating < player_div.min_rating:
+                reduction = 1
+            # Linear interpolation of score and rating to have players in top division sorted by rating
+            elif player_div.id == lowest_id and current_score.score < (player_div.highest_score - player_div.lowest_score) \
+                    * (rating - player_div.min_rating) / (player_div.max_rating - player_div.min_rating):
+                boost = 1
+
+            if outcome == "VICTORY":
+                current_score.score += 1 + boost
+            elif outcome == "DEFEAT":
+                current_score.score -= 1 + reduction
+            if current_score.score < 0:
+                current_score.score = 0
+
+            if current_score.score > player_div.highest_score:
+                if lowest_id < current_score.division_id:
+                    current_score.division_id -= 1
+            elif current_score.score < player_div.lowest_score:
+                if highest_id > current_score.division_id:
+                    current_score.division_id += 1
+
+        current_score.game_count += 1
+
         return current_score
+
+    @classmethod
+    def _get_player_division(cls, division_id, league):
```

Comment:
I think I would also add this as a method to `League`.
Something well-defined should happen if the `division_id` you are searching for is not found.
Either you throw an error here (preferred) or you make sure to guard against `None` returns when you are using it.

Reply:
How do I properly throw an error?

Reply:
```python

class CustomError(Exception):
  """Thrown if ..."""
  pass

def some_function():
  raise CustomError
```